File: e2e/graphql/guests_test.go
//go:build e2e

package graphql_test

import (
	"encoding/json"
	"net/http"
	"testing"
)

type guestSummary struct {
	ID        string `json:"id"`
	Name      string `json:"name"`
	Language  string `json:"language"`
	ExpiresAt string `json:"expiresAt"`
	CreatedAt string `json:"createdAt"`
}

func TestGuestDashboardAccessLifecycle(t *testing.T) {
	data, err := graphqlMutation(
		`mutation($input: CreateGuestInput!) {
			createGuest(input: $input) { id name language expiresAt createdAt }
		}`,
		map[string]any{"input": map[string]any{"name": "E2E Guest", "durationMinutes": 60, "language": "SV"}},
	)
	if err != nil {
		t.Fatalf("createGuest: %v", err)
	}
	var created struct {
		Guest guestSummary `json:"createGuest"`
	}
	if err := json.Unmarshal(data, &created); err != nil {
		t.Fatal(err)
	}
	t.Cleanup(func() {
		_, _ = graphqlMutation(`mutation($id: ID!) { deleteGuest(id: $id) }`, map[string]any{"id": created.Guest.ID})
	})

	login, status := rawPost(t, "",
		`mutation($name: String!) { guestLogin(name: $name) { token guest { id name language } } }`,
		map[string]any{"name": " e2e GUEST "},
	)
	if status != http.StatusOK || len(login.Errors) > 0 {
		t.Fatalf("guestLogin status=%d errors=%v", status, login.Errors)
	}
	var loggedIn struct {
		GuestLogin struct {
			Token string       `json:"token"`
			Guest guestSummary `json:"guest"`
		} `json:"guestLogin"`
	}
	if err := json.Unmarshal(login.Data, &loggedIn); err != nil {
		t.Fatal(err)
	}
	if loggedIn.GuestLogin.Token == "" || loggedIn.GuestLogin.Guest.ID != created.Guest.ID || loggedIn.GuestLogin.Guest.Language != "SV" {
		t.Fatalf("guest login = %+v", loggedIn.GuestLogin)
	}

	updated, _ := rawPost(t, loggedIn.GuestLogin.Token,
		`mutation { updateCurrentGuestLanguage(language: RU) { id language } }`, nil,
	)
	if len(updated.Errors) > 0 {
		t.Fatalf("update guest language: %v", updated.Errors)
	}

	dashboard, _ := rawPost(t, loggedIn.GuestLogin.Token,
		`query { currentGuest { id name language } devices { id } rooms { id } groups { id } scenes { id } dashboardLocalization { defaultContentLanguage } }`,
		nil,
	)
	if len(dashboard.Errors) > 0 {
		t.Fatalf("guest dashboard query: %v", dashboard.Errors)
	}
	adminData, _ := rawPost(t, loggedIn.GuestLogin.Token, `query { users { id } }`, nil)
	if len(adminData.Errors) == 0 {
		t.Fatal("guest could query users")
	}

	extended, err := graphqlMutation(
		`mutation($id: ID!) { extendGuest(id: $id, durationMinutes: 60) { id expiresAt } }`,
		map[string]any{"id": created.Guest.ID},
	)
	if err != nil || len(extended) == 0 {
		t.Fatalf("extendGuest: %v", err)
	}
	if _, err := graphqlMutation(`mutation($id: ID!) { deleteGuest(id: $id) }`, map[string]any{"id": created.Guest.ID}); err != nil {
		t.Fatalf("deleteGuest: %v", err)
	}
	afterRevoke, _ := rawPost(t, loggedIn.GuestLogin.Token, `query { devices { id } }`, nil)
	if len(afterRevoke.Errors) == 0 {
		t.Fatal("revoked guest could query the dashboard")
	}
}

func TestGuestDurationLimit(t *testing.T) {
	if err := graphqlMutationExpectError(
		`mutation($input: CreateGuestInput!) { createGuest(input: $input) { id } }`,
		map[string]any{"input": map[string]any{"name": "Too Long", "durationMinutes": 10081, "language": "EN"}},
	); err != nil {
		t.Fatalf("guest duration beyond seven days was accepted: %v", err)
	}
}

File: web/e2e/guests.test.ts
import { describe, expect, it } from "vitest";
import { graphql } from "$lib/gql";
import { getContext } from "./setup.js";

const CREATE_GUEST = graphql(`
  mutation E2ECreateGuest($input: CreateGuestInput!) {
    createGuest(input: $input) {
      id
      name
      expiresAt
      createdAt
    }
  }
`);

const EXTEND_GUEST = graphql(`
  mutation E2EExtendGuest($id: ID!) {
    extendGuest(id: $id, durationMinutes: 60) {
      id
      expiresAt
    }
  }
`);

const DELETE_GUEST = graphql(`
  mutation E2EDeleteGuest($id: ID!) {
    deleteGuest(id: $id)
  }
`);

interface RawGraphQLResponse {
  data?: Record<string, unknown>;
  errors?: Array<{ message: string; extensions?: { code?: string } }>;
}

async function rawGraphQL(
  token: string,
  query: string,
  variables = {},
): Promise<RawGraphQLResponse> {
  const { graphqlUrl } = getContext();
  const response = await fetch(graphqlUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify({ query, variables }),
  });
  return (await response.json()) as RawGraphQLResponse;
}

describe("guests", () => {
  it("limits a temporary guest to dashboard operations and revokes immediately", async () => {
    const { graphqlClient } = getContext();
    const created = await graphqlClient
      .mutation(CREATE_GUEST, { input: { name: "TS E2E Guest", durationMinutes: 60 } })
      .toPromise();
    expect(created.error).toBeUndefined();
    const guestId = created.data!.createGuest.id;

    try {
      const login = await rawGraphQL(
        "",
        `mutation($name: String!) { guestLogin(name: $name) { token guest { id name } } }`,
        { name: " ts e2e GUEST " },
      );
      expect(login.errors).toBeUndefined();
      const payload = login.data!.guestLogin as {
        token: string;
        guest: { id: string; name: string };
      };
      expect(payload.guest.id).toBe(guestId);

      const dashboard = await rawGraphQL(
        payload.token,
        `query { currentGuest { id } devices { id } rooms { id } groups { id } scenes { id } dashboardLocalization { defaultContentLanguage } }`,
      );
      expect(dashboard.errors).toBeUndefined();

      const users = await rawGraphQL(payload.token, `query { users { id } }`);
      expect(users.errors?.[0]?.extensions?.code).toBe("FORBIDDEN");

      const extended = await graphqlClient.mutation(EXTEND_GUEST, { id: guestId }).toPromise();
      expect(extended.error).toBeUndefined();
      const revoked = await graphqlClient.mutation(DELETE_GUEST, { id: guestId }).toPromise();
      expect(revoked.data?.deleteGuest).toBe(true);

      const afterRevoke = await rawGraphQL(payload.token, `query { devices { id } }`);
      expect(afterRevoke.errors?.[0]?.extensions?.code).toBe("UNAUTHENTICATED");
    } finally {
      await graphqlClient.mutation(DELETE_GUEST, { id: guestId }).toPromise();
    }
  });
});
